python.md

```diff
@@ -974,6 +974,49 @@ from path.to.some.module import \
     thing1, thing2, thing3, thing4
 ```
 
+### Make function signatures explicit
+
+Avoid using ``**kwargs`` in function signatures without good reason. Code with loosely defined
+function signatures can be difficult to work with, as it's unclear what variables are
+entering the function.
+ 
+```python
+def do_something(**kwargs):  # Don't do this
+   ...
+```
+
+Be explicit instead:
+
+```python
+def do_something(foo: int, bar: str):
+   ...
+```
+
+This includes functions that wrap lower level functionality, such as model creation methods:
+
+```python
+class MyModel(models.Model):
+    ...
+    
+    @classmethod
+    def new(cls, **kwargs):  # Don't do this.
+        return cls.objects.create(**kwargs)
+``` 
+
+Instead, do this:
+
+```python
+class MyModel(models.Model):
+    ...
+    
+    @classmethod
+    def new(cls, foo: int, bar: str):
```

Comment:
This suggests we can't use kwargs at all - I like this pattern which might be more relevant
```
def new(cls, *,  foo: int, bar: Optional[str] = None):
    ...
```

Reply:
Yeah, might be worth adding an example with a keyword argument, so that it's clear that you're anti-`**kwargs` not anti-*kwargs*. You could even replace 

> Avoid using `**kwargs`

– in your opening sentence with 

> Specify all the arguments (whether positional or keyword) you expect your function to take whenever possible. Avoid `*args`, `**kwargs`/`**params` etc.

Reply:
I think this is already covered by: https://github.com/octoenergy/conventions/blob/master/python.md#keyword-only-functions, do you agree?
It is related, though, maybe we should move that up so it follows this one.

Reply:
I don't think it does suggest that. It's always ok to use keyword arguments. The `func(..., *, ...)` construct forces them, but the absence of that construct doesn't discourage them.

_(Personally I'm not a fan of `func(..., *, ...)` but that's a separate issue)_

Reply:
Re. @timwraight comment, could we lean on the language used by the Python glossary?

> _var-keyword_: specifies that arbitrarily many keyword arguments can be provided (in addition to any keyword arguments already accepted by other parameters). Such a parameter can be defined by prepending the parameter name with `**`, for example _kwargs_ in the example above.


https://docs.python.org/3/glossary.html#term-parameter

Reply:
Thanks all. I have used @timwraight and @jamesbeith's suggestions in https://github.com/octoenergy/conventions/pull/66/commits/66f5f8ea5b3cda151a19c97b8fa4490bf8915194.